### PR TITLE
Part 5: tests

### PR DESCRIPTION
**Pull request by apostasie**
Broken out of #3535

### CODE REVIEW

File: cmd/nerdctl/ipfs/ipfs_compose_linux_test.go
Comment (AkihiroSuda) [inline cmd/nerdctl/ipfs/ipfs_compose_linux_test.go:156]: What is "this"? The volume?

Comment (apostasie) [reply cmd/nerdctl/ipfs/ipfs_compose_linux_test.go:156]: Yes.

Comment (apostasie) [reply cmd/nerdctl/ipfs/ipfs_compose_linux_test.go:156]: On occasions php seem to hard fail trying to read a file from the volume that it is somehow expecting.

Comment (apostasie) [reply cmd/nerdctl/ipfs/ipfs_compose_linux_test.go:156]: Curious if the compose global lock PR will help with these.

File: pkg/testutil/testutil.go
Comment (AkihiroSuda) [inline pkg/testutil/testutil.go:791]: This sounds like overestimating 🤔

Now those tests are skipped on the CI?

Comment (apostasie) [reply pkg/testutil/testutil.go:791]: Nothing changes for the pre-existing tests - they will still be retried, and will not be skipped.

The difference is that the new tests now run first, in a separate step of the job, without retries.

Comment (apostasie) [reply pkg/testutil/testutil.go:791]: `./test-integration.sh -test.only-flaky`:
- does run with retries, and includes all legacy tests, and all new tests that have been flagged flaky explicitly

`./test-integration.sh`:
- default behavior, runs only the non-flaky tests

The CI runs both, as successive steps of the test job.

Comment (apostasie) [reply pkg/testutil/testutil.go:791]: And yes, we need the github actions PR once this one will be merged.

Comment (AkihiroSuda) [reply pkg/testutil/testutil.go:791]: What about adding a flag like `-test.skip-legacy` instead?

Not every legacy test should be assumed to be flaky?

File: pkg/testutil/test/helpers.go
Comment (apostasie) [reply pkg/testutil/test/helpers.go:73]: 1 ensure the task fails and exitcode == 1.

-1 ensure the task fails but does not check what the exit code is

Comment (apostasie) [reply pkg/testutil/test/helpers.go:73]: https://github.com/containerd/nerdctl/blob/main/docs/testing/tools.md#expectations

File: cmd/nerdctl/container/container_commit_linux_test.go
Comment (apostasie) [reply cmd/nerdctl/container/container_commit_linux_test.go:50]: Yes.
https://github.com/containerd/nerdctl/blob/main/docs/testing/tools.md#setup-and-cleanup

File: pkg/testutil/testutil.go
Comment (apostasie) [reply pkg/testutil/testutil.go:791]: > What about adding a flag like `-test.skip-legacy` instead?
> 
> Not every legacy test should be assumed to be flaky?

But this is what we are doing right now: any existing is treated as flaky and will be retried if it fails.

This PR does not change anything for that.

File: cmd/nerdctl/container/container_commit_linux_test.go
Comment (apostasie) [reply cmd/nerdctl/container/container_commit_linux_test.go:45]: Not necessary per-se, but what asked for when the k8s env was introduced: https://github.com/containerd/nerdctl/pull/3296#discussion_r1719194632

File: pkg/testutil/testutil.go
Comment (apostasie) [reply pkg/testutil/testutil.go:791]: > > What about adding a flag like `-test.skip-legacy` instead?
> > Not every legacy test should be assumed to be flaky?
> 
> But this is what we are doing right now: any existing test is treated as flaky and will be retried if it fails.
> 
> This PR does not change anything for that.

Current situation:
- every test is treated as flaky and will be retried multiple times until it passes

With these two PRs:
- pre-existing tests and new tests marked "isflaky" are treated exactly the same as previously and will be retried if they fail - this is what is called "flaky"
- new tests are only run once and are not retried - this is what is called "non-flaky"

Hope that clarifies.

Comment (djdongjin) [reply pkg/testutil/testutil.go:791]: How do we distinguish between a `pre-existing` tests and `new` tests?

Do we have many flaky tests? If not I'd prefer assuming all tests (regardless new or pre-existing) are not flaky, and mark individual ones as flaky explicitly.

Comment (apostasie) [reply pkg/testutil/testutil.go:791]: > Do we have many flaky tests? 

As stated previously: we currently re-run all tests multiple times if they fail, effectively treating _every_ test as flaky.
This is the current situation, which is obviously hiding all errors.

I tried multiple times to disable retries before:
- https://github.com/containerd/nerdctl/issues/3303
- https://github.com/containerd/nerdctl/pull/3400
- https://github.com/containerd/nerdctl/pull/3189

> If not I'd prefer assuming all tests (regardless new or pre-existing) are not flaky, and mark individual ones as flaky explicitly.

I do not see how we can do that _from the current situation_, which is as stated to treat  everything as flaky.

Comment (apostasie) [reply pkg/testutil/testutil.go:791]: > Do we have many flaky tests? If not I'd prefer assuming all tests (regardless new or pre-existing) are not flaky, and mark individual ones as flaky explicitly.

I offered this before in https://github.com/containerd/nerdctl/issues/3303 

That would mean the CI would be very finicky and failing for weeks (or months, depending on time invested in this) and we would have to go through a lot of pain to get it back to any form of green.

Furthermore - and this is one of the reasons I started the test rewrite effort, a significant fraction of the failures are problems with the test, not with the code.

At least with the rewrite, I know that the problems are with the code.

Hope that clarifies.

File: cmd/nerdctl/container/container_commit_linux_test.go
Comment (djdongjin) [reply cmd/nerdctl/container/container_commit_linux_test.go:45]: > Not necessary per-se

will these commands be run in the default namespace or in `nerdctl-test-k8s`?

Comment (apostasie) [reply cmd/nerdctl/container/container_commit_linux_test.go:45]: k8s namespaces are not the same as containerd namespaces (yeah, really confusing).

Commands run by `nerdctl` for _this environment_ are run inside the `k8s.io` containerd namespace (instead of the `nerdctl-test` usual namespace).

Then `nerdctl-test-k8s` is a k8 namespace, unrelated.

Also note that these are running against the containerd provided by kind, which is inside a container IIRC, and not on the host.

**Comment by apostasie** [comment]
@djdongjin I am commenting out this with the latest push https://github.com/containerd/nerdctl/compare/92b6ebc138b6d7258692b0f72541a90ba00714c4..6ddb9ec0a700068bbc2233bf011a1184a1a8c1e2

Otherwise, must-retry-tests will not be run until we merge #3535 (which this PR has been broken out of by maintainers request).

Will uncomment again for #3535

**Comment by apostasie** [comment]
Rebased.

**Comment by apostasie** [comment]
Folks, breaking out some changes in separate PRs:
- #3565 
- #3564

Will rebase on merge.

**Comment by apostasie** [comment]
Rebased.

Marking legacy tests as flaky will be done in #3535 instead.

We are left here with kube tests and tooling cleanups, registry tooling cleanup, and "commit" tests rewrite.

**Comment by apostasie** [comment]
Rebased.

File: cmd/nerdctl/container/container_commit_linux_test.go
Comment (AkihiroSuda) [inline cmd/nerdctl/container/container_commit_linux_test.go:37]: nerdctl-test-k8s should a constant

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Comment by apostasie** [comment]
Thanks @djdongjin and @AkihiroSuda .

**Pull request merged**
